vnext/Microsoft.ReactNative.Cxx/JSI/JsiAbiApi.cpp

```diff
@@ -241,31 +241,75 @@ operator()(JsiRuntime const &runtime, JsiValueRef const &thisArg, array_view<Jsi
   VerifyElseCrashSz(false, "Function is not a host function");
 }
 
+//===========================================================================
+// JsiAbiRuntimeHolder implementation
+//===========================================================================
+
+JsiAbiRuntimeHolder::JsiAbiRuntimeHolder(std::unique_ptr<JsiAbiRuntime> jsiRuntime) noexcept
+    : m_jsiRuntime{std::move(jsiRuntime)}, m_isOwning{true} {}
+
+JsiAbiRuntimeHolder::JsiAbiRuntimeHolder(JsiAbiRuntime *jsiRuntime) noexcept : m_jsiRuntime{jsiRuntime} {}
+
+JsiAbiRuntimeHolder::operator bool() noexcept {
+  return m_jsiRuntime != nullptr;
+}
+
+JsiAbiRuntime &JsiAbiRuntimeHolder::operator*() noexcept {
+  return *m_jsiRuntime;
+}
+
+JsiAbiRuntime *JsiAbiRuntimeHolder::operator->() noexcept {
+  return m_jsiRuntime.get();
+}
+
+JsiAbiRuntimeHolder ::~JsiAbiRuntimeHolder() noexcept {
+  if (!m_isOwning) {
+    m_jsiRuntime.release();
+  }
+}
+
 //===========================================================================
 // JsiAbiRuntime implementation
 //===========================================================================
 
-/*static*/ std::mutex JsiAbiRuntime::s_mutex;
-/*static*/ std::map<void *, JsiAbiRuntime *> JsiAbiRuntime::s_jsiAbiRuntimeMap;
+static thread_local std::map<void *, JsiAbiRuntime *> *tls_jsiAbiRuntimeMap{nullptr};
 
 JsiAbiRuntime::JsiAbiRuntime(JsiRuntime const &runtime) noexcept : m_runtime{runtime} {
-  std::scoped_lock lock{s_mutex};
-  s_jsiAbiRuntimeMap.try_emplace(get_abi(runtime), this);
+  VerifyElseCrashSz(
+      GetFromJsiRuntime(runtime) == nullptr,
+      "We can have only one instance of JsiAbiRuntime for JsiRuntime in the thread.");
+  if (!tls_jsiAbiRuntimeMap) {
+    tls_jsiAbiRuntimeMap = new std::map<void *, JsiAbiRuntime *>();
+  }
+  tls_jsiAbiRuntimeMap->try_emplace(get_abi(runtime), this);
 }
 
 JsiAbiRuntime::~JsiAbiRuntime() {
-  std::scoped_lock lock{s_mutex};
-  s_jsiAbiRuntimeMap.erase(get_abi(m_runtime));
+  VerifyElseCrashSz(
+      GetFromJsiRuntime(m_runtime) != nullptr, "JsiAbiRuntime must be called in the same thread where it was created.");
+  tls_jsiAbiRuntimeMap->erase(get_abi(m_runtime));
+  if (tls_jsiAbiRuntimeMap->empty()) {
+    delete tls_jsiAbiRuntimeMap;
+    tls_jsiAbiRuntimeMap = nullptr;
+  }
 }
 
-/*static*/ JsiAbiRuntime *JsiAbiRuntime::FromJsiRuntime(JsiRuntime const &jsiRuntime) noexcept {
-  std::scoped_lock lock{s_mutex};
-  auto it = s_jsiAbiRuntimeMap.find(get_abi(jsiRuntime));
-  if (it != s_jsiAbiRuntimeMap.end()) {
-    return it->second;
-  } else {
-    return nullptr;
+/*static*/ JsiAbiRuntime *JsiAbiRuntime::GetFromJsiRuntime(JsiRuntime const &runtime) noexcept {
+  if (tls_jsiAbiRuntimeMap && runtime) {
+    auto it = tls_jsiAbiRuntimeMap->find(get_abi(runtime));
+    if (it != tls_jsiAbiRuntimeMap->end()) {
+      return it->second;
+    }
+  }
+  return nullptr;
```

Comment:
Should we log this case? I'm imagining a threading error when this code runs on a different thread than the initializer?
Totally fine if that is highly unlikley to happen.

Reply:
In GetFromJsiRuntime function it should be OK to return null because there can be no associations which is a legal case. I do not think we need to log this this because nullptr is one of the expected values here.